crypto-ffi/src/wasm/mod.rs

```diff
@@ -196,46 +203,147 @@ pub(crate) enum InternalError {
+        }
+
+        // What now? We only really care about the innermost variants, not the error stack, but that produces
+        // an arbitrary set of types. We can't match against that!
+        //
+        // Or at least, not without the power of macros. We can use them to match against heterogenous types.
+
+        /// Like [`matches!`], but with an out expression which can reference items captured by the pattern.
+        ///
+        /// Hopefully only ever use this in conjunction with `interior_matches!`, because for most sane
+        /// circumstances, `if let` is the better design pattern.
+        macro_rules! matches_option {
+            ($val:expr, $pattern:pat $(if $guard:expr)? => $out:expr) => {
+                match ($val) {
+                    $pattern $(if $guard)? => Some($out),
+                    _ => None,
+                }
+            };
+        }
+
+        /// This is moderately horrific and we hopefully will not require it anywhere else, but
+        /// it solves a real problem here: how do we match against the innermost error variants,
+        /// when we have a heterogenous set of types to match against?
+        macro_rules! match_heterogenous {
+            ($err:expr => {
+                $( $pattern:pat $(if $guard:expr)? => $var:expr, )*
+                ||=> $default:expr,
+            }) => {{
+                if false {unreachable!()}
+                $(
+                    else if let Some(v) = matches_option!($err.downcast_ref(), Some($pattern) $(if $guard)? => $var) {
+                        v
+                    }
+                )*
+                else {
+                    $default
+                }
+            }};
+        }
+
+        match_heterogenous!(innermost => {
+            core_crypto::LeafError::ConversationAlreadyExists(id) => MlsError::ConversationAlreadyExists(id.clone()).into(),
+            core_crypto::mls::conversation::Error::BufferedFutureMessage => MlsError::BufferedFutureMessage.into(),
+            core_crypto::mls::conversation::Error::DuplicateMessage => MlsError::DuplicateMessage.into(),
+            core_crypto::mls::conversation::Error::MessageEpochTooOld => MlsError::MessageEpochTooOld.into(),
+            core_crypto::mls::conversation::Error::SelfCommitIgnored => MlsError::SelfCommitIgnored.into(),
+            core_crypto::mls::conversation::Error::StaleCommit => MlsError::StaleCommit.into(),
+            core_crypto::mls::conversation::Error::StaleProposal => MlsError::StaleProposal.into(),
+            core_crypto::mls::conversation::Error::UnbufferedFarFutureMessage => MlsError::WrongEpoch.into(),
+            core_crypto::mls::conversation::Error::OrphanWelcome => MlsError::OrphanWelcome.into(),
+            core_crypto::mls::Error::UnmergedPendingGroup => MlsError::UnmergedPendingGroup.into(),
+            ||=> MlsError::Other(error.innermost_error_message()).into(),
+        })
+    }
```

Comment:
Looks like this entire code block is shared between wasm and uniffi bindings.
Can we de-duplicate it somehow?

Reply:
I mean, massive parts of the code are shared between wasm and uniffi bindings. The right way to deduplicate this is to unify those bindings and use conditional compilation to expose any differences we require. 

Deduplicating these macros means either putting them in `core-crypto`, or into some other crate which is parent to both `generic` and `wasm`. But that involves making them public. I'm happier if they stay limited to the context of this one very specific function.